---
title: "Applying spatialDE to identify spatially DE genes"
author: Stephanie Hicks
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries
```{r}
suppressMessages({
    library(here)
    library(SingleCellExperiment)
})
```

Copy data (if needed) to be able to work with the here package.
```{r}
if(!dir.exists(here("sample_data"))){
    dir.create(here("sample_data"))
    processedData_sce_scran <- "/dcl02/lieber/ajaffe/SpatialTranscriptomics/HumanPilot/Analysis/Human_DLPFC_Visium_processedData_sce_scran.Rdata"
    file.copy(processedData_sce_scran, here("sample_data"))
}
```

Pick one sample (e.g. `sample_name==151673`) and create a dense count matrix and meta data `.csv` filel to use as input into `spatialDE`. 
```{r}
load(here("sample_data", "Human_DLPFC_Visium_processedData_sce_scran.Rdata"))

# let's try first just picking one image
ix <- colData(sce)$sample_name == 151673
sce <- sce[, ix]

pryr::object_size(counts(sce)) # 99 MB (using sparse matrices)
tmp <- t(as.matrix(counts(sce)))
pryr::object_size(tmp) # not using sparse matrices
write.csv(tmp, quote = FALSE, row.names = TRUE,
          file = here("sample_data", "Human_DLPFC_Visium_processedData_sce_scran_counts.csv"))

is.mito <- grep("^MT-", rowData(sce)$gene_name)
df <- scater::perCellQCMetrics(sce, subsets=list(Mito=is.mito))
df <- cbind(colData(sce), df)
head(df)

write.csv(df, quote = FALSE, row.names = TRUE,
          file = here("sample_data", "Human_DLPFC_Visium_processedData_sce_scran_meta.csv"))
```

This code is the same as above, but I have extended it to create dense count matrices and meta data files for all twelves samples. 
```{r}
if(!dir.exists(here("sample_data", "by_sample_id"))){
    dir.create(here("sample_data", "by_sample_id"))
    
    # create individual sample folders
    samp_ids <- names(table(colData(sce)$sample_name))
    for(samp in samp_ids[-1]){
        ix <- colData(sce)$sample_name == samp
        samp_name_counts <- paste("sample", samp, "counts.csv", sep = "_")
        samp_name_meta <- paste("sample", samp, "meta.csv", sep = "_")
        sce_sub <- sce[, ix]

        tmp <- t(as.matrix(counts(sce_sub)))
        if(!file.exists(here("sample_data", "by_sample_id", samp_name_counts))){
            write.csv(tmp, quote = FALSE, row.names = TRUE,
                      file = here("sample_data", "by_sample_id", samp_name_counts))
        }

        is.mito <- grep("^MT-", rowData(sce_sub)$gene_name)
        df <- scater::perCellQCMetrics(sce_sub, subsets=list(Mito=is.mito))
        df <- cbind(colData(sce_sub), df)

        if(!file.exists(here("sample_data", "by_sample_id", samp_name_meta))){
            write.csv(df, quote = FALSE, row.names = TRUE,
                      file = here("sample_data", "by_sample_id", samp_name_meta))
        }
        rm(sce_sub, tmp, is.mito, df)
    }
   
 }

```


# Spatial DE 

## Getting python 3 set up on jhpce

Here I am using Python 3. 

I had to upate the numpy modules and install the scipy and pandas modules

```{bash}
module load conda_R/devel
python3 -m pip install numpy --upgrade --user
python3 -m pip install scipy --user
python3 -m pip install pandas --user
python3 -m pip install patsy --user
python3 -m pip install matplotlib --user
```

## installing `SpatialDE` and `NaiveDE` Python modules

To install the [`SpatialDE`](https://github.com/Teichlab/SpatialDE) 
and [`NaiveDE`](https://github.com/Teichlab/NaiveDE) packages, use 

```{bash}
git clone git@github.com:Teichlab/SpatialDE.git
python3 -m pip install SpatialDE --user

git clone git@github.com:Teichlab/NaiveDE.git
python3 -m pip install NaiveDE --user
```

Python modules should be install `/users/<username>/.local/lib/python3.8/site-packages`

```{r}
library(reticulate)
use_python("/jhpce/shared/jhpce/core/conda/miniconda3-4.6.14/envs/svnR-devel/bin/python3")
repl_python()
```

Use `repl_python()` python repl from the R command. Use `exit` to exit the repl.
**Note**: Objects do not have permenancy in R after exiting the python repl.
```{python}
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import datetime

mpl.rcParams['axes.spines.right'] = False
mpl.rcParams['axes.spines.top'] = False

import NaiveDE
import SpatialDE
```

## Applying `SpatialDE` for one sample 

Load count data
```{python}
counts = pd.read_csv('/fastscratch/myscratch/shicks1/HumanPilot/sample_data/Human_DLPFC_Visium_processedData_sce_scran_counts.csv', index_col=0)
counts.shape # get data frame dimension
counts = counts.T[counts.sum(axis=0) >= 5].T  # Filter practically unobserved genes

print(counts.shape)
counts.iloc[:5, :5]
```

Load spatial coordinates
```{python}
sample_info = pd.read_csv('/fastscratch/myscratch/shicks1/HumanPilot/sample_data/Human_DLPFC_Visium_processedData_sce_scran_meta.csv', index_col=0)
sample_info.shape
```

```{python}
figsize(6, 4)
plt.scatter(sample_info['imagerow'], sample_info['imagecol'], c='k');
plt.axis('equal');
plt.show()
```

Authors of spatialDE recommend using linear regression to remove technical variation 
due to library size or sequencing depth on spatial samples (will bias expression in 
every gene otherwise) before performing the spatial test.

```{python}
norm_expr = NaiveDE.stabilize(counts.T).T
resid_expr = NaiveDE.regress_out(sample_info, norm_expr.T, 'np.log(sum)').T
```

For the sake of this example, let's just run the test on 1000 random genes. This should just take a few seconds. With our very fast implementation, testing all 14,000 genes takes about 10 minutes.
```{python}
# downsample to 1500 spots
sample_resid_expr = resid_expr.sample(n=1500, axis=0, random_state=1)
sample_sample_info = sample_info.loc[sample_resid_expr.index]  # Align sample_info matrix with reduced counts table

# sample 100 genes
sample_resid_expr = sample_resid_expr.sample(n=100, axis=1, random_state=1)

X = sample_sample_info[['imagerow', 'imagecol']]

print(datetime.datetime.now())
results = SpatialDE.run(X, sample_resid_expr)
print(datetime.datetime.now())

# Save spatial results
results.to_csv('/fastscratch/myscratch/shicks1/HumanPilot/sample_data/Human_DLPFC_Visium_processedData_sce_scran_spatialDE_results.csv')

de_results = results[(results.qval < 0.05)].copy()
ms_results = SpatialDE.model_search(X, sample_resid_expr, de_results)

ms_results.to_csv('/fastscratch/myscratch/shicks1/HumanPilot/sample_data/Human_DLPFC_Visium_processedData_sce_scran_spatialDE_MS_results.csv')
```




## Applying `SpatialDE` for all 12 samples

```{python}
import glob
samp_counts = glob.glob('/fastscratch/myscratch/shicks1/HumanPilot/sample_data/by_sample_id/*_counts.csv')
samp_meta = [[x[:-10] + 'meta.csv'] for x in samp_counts]
samp_output = [[x[:-10] + 'spatialDE_results.csv'] for x in samp_counts]
df = pd.DataFrame({'counts':samp_counts, 'meta':samp_meta, 'out':samp_output})

# make sure I remember for loops in python and the order of the samples matches
for index, row in df.iterrows():
    print(row['counts'], row['meta'], row['out'])

# loop over all twelve samples and save output
for index, row in df.iterrows():
    counts = pd.read_csv(row['counts'], index_col=0) # load counts
    counts = counts.T[counts.sum(axis=0) >= 5].T  # Filter practically unobserved genes
    sample_info = pd.read_csv(row['meta'], index_col=0) # load meta data with spatial coordinates
    norm_expr = NaiveDE.stabilize(counts.T).T # remove tech variation
    resid_expr = NaiveDE.regress_out(sample_info, norm_expr.T, 'np.log(sum)').T
    sample_resid_expr = resid_expr.sample(n=1500, axis=0, random_state=1) # downsample to 1500 spots
    sample_sample_info = sample_info.loc[sample_resid_expr.index]  # Align sample_info matrix with reduced counts table
    X = sample_sample_info[['imagerow', 'imagecol']]
    results = SpatialDE.run(X, sample_resid_expr)
    results.to_csv(row['out']) # Save spatial results

```


